File: backend/app/controllers/user_controller.py
import hashlib
from datetime import datetime, timezone

from bson import ObjectId
from fastapi import HTTPException, status

from backend.app.db.mongo import get_database
from backend.app.models.user_model import (
    UserCreateRequest,
    UserLoginRequest,
    UserUpdateRequest,
)


def _hash_password(password: str) -> str:
    return hashlib.sha256(password.encode("utf-8")).hexdigest()


def _normalize_email(email: str) -> str:
    return email.strip().lower()


def _serialize_user(document: dict) -> dict:
    return {
        "id": str(document["_id"]),
        "name": document["name"],
        "email": document["email"],
        "joined_at": document["created_at"].isoformat(),
    }


def _to_object_id(id_value: str) -> ObjectId:
    try:
        return ObjectId(id_value)
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid user id format.",
        ) from exc


async def register_user(payload: UserCreateRequest) -> dict:
    db = get_database()
    users = db.users

    normalized_email = _normalize_email(payload.email)
    existing = await users.find_one({"email": normalized_email})
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="User with this email already exists.",
        )

    now = datetime.now(timezone.utc)
    document = {
        "name": payload.name.strip(),
        "email": normalized_email,
        "password_hash": _hash_password(payload.password),
        "created_at": now,
        "updated_at": now,
    }

    result = await users.insert_one(document)
    created_user = await users.find_one({"_id": result.inserted_id})
    return _serialize_user(created_user)


async def login_user(payload: UserLoginRequest) -> dict:
    db = get_database()
    users = db.users

    normalized_email = _normalize_email(payload.email)
    user = await users.find_one(
        {
            "email": normalized_email,
            "password_hash": _hash_password(payload.password),
        }
    )

    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid email or password.",
        )

    return _serialize_user(user)


async def get_user_by_id(user_id: str) -> dict:
    db = get_database()
    users = db.users

    oid = _to_object_id(user_id)
    user = await users.find_one({"_id": oid})
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found.",
        )

    return _serialize_user(user)


async def update_user(user_id: str, payload: UserUpdateRequest) -> dict:
    db = get_database()
    users = db.users

    oid = _to_object_id(user_id)
    user = await users.find_one({"_id": oid})
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found.",
        )

    update_doc = {}

    if payload.name is not None:
        update_doc["name"] = payload.name.strip()

    if payload.email is not None:
        normalized_email = _normalize_email(payload.email)
        existing = await users.find_one({"email": normalized_email, "_id": {"$ne": oid}})
        if existing:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Another user already uses this email.",
            )
        update_doc["email"] = normalized_email

    if payload.password is not None:
        update_doc["password_hash"] = _hash_password(payload.password)

    if not update_doc:
        return _serialize_user(user)

    update_doc["updated_at"] = datetime.now(timezone.utc)

    await users.update_one({"_id": oid}, {"$set": update_doc})
    updated_user = await users.find_one({"_id": oid})
    return _serialize_user(updated_user)

File: Web_Backend/controllers/userController.js
const User = require("../models/User");

const EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

const serializeUser = (userDoc) => ({
  id: userDoc._id.toString(),
  name: userDoc.name,
  email: userDoc.email,
  joined_at: userDoc.createdAt.toISOString(),
});

const validateRegisterBody = ({ name, email, password }) => {
  if (!name || !email || !password) {
    return "Name, email, and password are required.";
  }
  if (!EMAIL_REGEX.test(String(email).toLowerCase())) {
    return "Invalid email format.";
  }
  if (String(password).length < 6) {
    return "Password must be at least 6 characters.";
  }
  return null;
};

const registerUser = async (req, res) => {
  try {
    const validationError = validateRegisterBody(req.body);
    if (validationError) {
      return res.status(400).json({ detail: validationError });
    }

    const { name, email, password } = req.body;

    const existingUser = await User.findOne({ email: email.toLowerCase().trim() });
    if (existingUser) {
      return res.status(409).json({ detail: "User with this email already exists." });
    }

    const user = await User.create({
      name: String(name).trim(),
      email: String(email).toLowerCase().trim(),
      password: String(password),
    });

    return res.status(201).json(serializeUser(user));
  } catch (error) {
    if (error && error.code === 11000) {
      return res.status(409).json({ detail: "User with this email already exists." });
    }
    return res.status(500).json({ detail: "Failed to register user." });
  }
};

const loginUser = async (req, res) => {
  try {
    const { email, password } = req.body;
    if (!email || !password) {
      return res.status(400).json({ detail: "Email and password are required." });
    }

    const normalizedEmail = String(email).toLowerCase().trim();
    const user = await User.findOne({ email: normalizedEmail });

    if (!user) {
      return res.status(401).json({ detail: "Invalid email or password." });
    }

    const isPasswordValid = await user.comparePassword(String(password));
    if (!isPasswordValid) {
      return res.status(401).json({ detail: "Invalid email or password." });
    }

    return res.status(200).json(serializeUser(user));
  } catch (error) {
    return res.status(500).json({ detail: "Failed to login user." });
  }
};

const getUserById = async (req, res) => {
  try {
    const { userId } = req.params;
    const user = await User.findById(userId);

    if (!user) {
      return res.status(404).json({ detail: "User not found." });
    }

    return res.status(200).json(serializeUser(user));
  } catch (error) {
    return res.status(400).json({ detail: "Invalid user id." });
  }
};

const updateUser = async (req, res) => {
  try {
    const { userId } = req.params;
    const { name, email, password } = req.body;

    const user = await User.findById(userId);
    if (!user) {
      return res.status(404).json({ detail: "User not found." });
    }

    if (name !== undefined) {
      if (!String(name).trim()) {
        return res.status(400).json({ detail: "Name cannot be empty." });
      }
      user.name = String(name).trim();
    }

    if (email !== undefined) {
      const normalizedEmail = String(email).toLowerCase().trim();
      if (!EMAIL_REGEX.test(normalizedEmail)) {
        return res.status(400).json({ detail: "Invalid email format." });
      }

      const existing = await User.findOne({
        email: normalizedEmail,
        _id: { $ne: user._id },
      });

      if (existing) {
        return res.status(409).json({ detail: "Another user already uses this email." });
      }

      user.email = normalizedEmail;
    }

    if (password !== undefined) {
      if (String(password).length < 6) {
        return res.status(400).json({ detail: "Password must be at least 6 characters." });
      }
      user.password = String(password);
    }

    await user.save();

    return res.status(200).json(serializeUser(user));
  } catch (error) {
    if (error && error.code === 11000) {
      return res.status(409).json({ detail: "Another user already uses this email." });
    }
    return res.status(400).json({ detail: "Failed to update user." });
  }
};

module.exports = {
  registerUser,
  loginUser,
  getUserById,
  updateUser,
};

